### PR TITLE
[cmake] Refactor build

### PR DESCRIPTION
**Pull request by xoofx**
This is a split of the previous PR #68 for 1) and 2)

- Remove usage of external `cmake -E` executable as they were not failsafe
  (not detecting errors), and use `add_custom_command` instead
- move all mkrename options to Configure.cmake
- fix AVX512F typo and compilation detection

### CODE REVIEW

**Comment by xoofx** [comment]
I will add the `appveyor.yml` as part of another PR once this one is merged (as it may require a few iterations to check that it is working well)

**Comment by xoofx** [comment]
Ok, build is green

File: src/libm/CMakeLists.txt [hunk 1]
```diff
@@ -99,48 +141,48 @@ set_target_properties(${TARGET_OBJECTS} ${TARGET_OBJECTS_GNUABI} PROPERTIES
...
+
+foreach(SIMD ${SLEEF_HEADER_LIST})
+  add_custom_command(TARGET ${TARGET_HEADERS} 
+    COMMAND echo Generating sleef.h: ${TARGET_MKRENAME} ${HEADER_PARAMS_${SIMD}}
+    COMMAND ${TARGET_MKRENAME} ${HEADER_PARAMS_${SIMD}} >> ${SLEEF_GEN_HEADER}
+  )
+endforeach()
+
+file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/sleef_footer.h "#undef IMPORT\n#endif\n")
+if(MSVC)
```
Comment (fpetrogalli-arm) [inline src/libm/CMakeLists.txt]: This is also an MSVC change that I think shound't be here. Please remove this if statement (only the true branch!! :))

File: src/libm/CMakeLists.txt [hunk 2]
```diff
@@ -99,48 +141,48 @@ set_target_properties(${TARGET_OBJECTS} ${TARGET_OBJECTS_GNUABI} PROPERTIES
...
+  )
+endif()
+
+
 
 # Install libsleef and sleef.h
 install(FILES ${SLEEF_INCLUDE_HEADER}
   DESTINATION include)
-install(TARGETS ${TARGET_LIBSLEEF} ${TARGET_LIBSLEEFGNUABI}
+if(SLEEF_ENABLE_GNUABI)
```
Comment (fpetrogalli-arm) [inline src/libm/CMakeLists.txt]: We want to install `libsleef` unconditionally. Please change this to:

```
if(GNUABI)
  install(libsleefgnuabi)
endif()

install (libsleef)
```

Comment (xoofx) [reply src/libm/CMakeLists.txt]: Fixed by push+squash on 0de75b5

File: src/libm/CMakeLists.txt (hunk 1 above)
Comment (xoofx) [reply src/libm/CMakeLists.txt]: Fixed by push+squash on 0de75b5

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -2,5 +2,10 @@ include_directories("common")
 include_directories("arch")
 
 add_subdirectory("libm")
-add_subdirectory("libm-tester")
+# We might not have MPFR installed, so skipt it if not found instead of crashing
+if(LIB_MPFR)
```
[the remark this answers is not in the document]
Comment (xoofx) [reply src/CMakeLists.txt]: Fixed by push+squash on 0de75b5

File: Configure.cmake [hunk 4]
```diff
@@ -19,43 +22,114 @@ if(NOT CMAKE_BUILD_TYPE)
...
-  # Intel vector extensions.
-  set(FLAGS_ENABLE_SSE2 "-msse2")
-  set(FLAGS_ENABLE_SSE4 "-msse4.1")
-  set(FLAGS_ENABLE_AVX "-mavx")
-  set(FLAGS_ENABLE_FMA4 "-mfma4")
-  set(FLAGS_ENABLE_AVX2 "-mavx2;-mfma")
-  set(FLAGS_ENABLE_AVX2128 "-mavx2;-mfma")
-  set(FLAGS_ENABLE_AVX512F "-mavx512f")
-
+elseif(MSVC)
```
[the remark this answers is not in the document]
Comment (xoofx) [reply Configure.cmake]: Fixed by push+squash on 0de75b5

File: Configure.cmake [hunk 5]
```diff
@@ -19,43 +22,114 @@ if(NOT CMAKE_BUILD_TYPE)
...
+
+command_arguments(RENAME_PARAMS_GNUABI_SSE2    sse2 b 2 4 _mm128d _mm128 _mm128i _mm128i __SSE2__)
+command_arguments(RENAME_PARAMS_GNUABI_AVX     avx c 4 8 __m256d __m256 __m128i "struct { __m128i x, y$<SEMICOLON> }" __AVX__)
+command_arguments(RENAME_PARAMS_GNUABI_AVX2    avx2 d 4 8 __m256d __m256 __m128i __m256i __AVX2__)
+command_arguments(RENAME_PARAMS_GNUABI_AVX512F avx512f e 8 16 __m512d __m512 __m256i __m512i __AVX512F__)
+command_arguments(RENAME_PARAMS_GNUABI_ADVSIMD advsimd n 2 4 float64x2_t float32x4_t int32x2_t int32x4_t __ARM_NEON)
+
 # COMPILER DETECTION
 
+# Allow to define the Gcc/Clang here
```
[the remark this answers is not in the document]
Comment (xoofx) [reply Configure.cmake]: Fixed by push+squash on 0de75b5

File: Configure.cmake [hunk 6]
```diff
@@ -19,43 +22,114 @@ if(NOT CMAKE_BUILD_TYPE)
...
+    ADVSIMD
+  )
+  set(HEADER_PARAMS_ADVSIMD    "2 4 float64x2_t float32x4_t int32x2_t int32x4_t __ARM_NEON advsimd")
 
-if(SLEEF_ARCH_X86 OR SLEEF_ARCH_AARCH64)
-  set(SLEEF_ENABLE_GNUABI ON CACHE INTERNAL "Build GNU ABI compatible version.")
 endif()
 
+# Enable building of the GNU ABI version for x86 and aarch64
+if(NOT MSVC)
```
[the remark this answers is not in the document]
Comment (xoofx) [reply Configure.cmake]: Fixed by push+squash on 0de75b5

File: Configure.cmake [hunk 7]
```diff
@@ -2,9 +2,12 @@ include(CheckCCompilerFlag)
 include(CheckCSourceCompiles)
 include(CheckTypeSize)
 
+# Check that the library LIB_MPFR is available
+find_library(LIB_MPFR mpfr)
+
 # The library currently supports the following SIMD architectures
 set(SLEEF_SUPPORTED_EXTENSIONS
-  SSE2 SSE4 AVX FMA4 AVX2 AVX2128 AVX512F # x86
+  SSE2 SSE4 AVX FMA4 AVX2 AVX512F AVX2128  # x86
```
[the remark this answers is not in the document]
Comment (xoofx) [reply Configure.cmake]: Fixed by push+squash on 0de75b5

**Comment by xoofx** [comment]
Repushed everything as part of commit f94cf2a

**Comment by xoofx** [comment]
Errata consider 4192136 (didn't check that it was compiling,, as I'm doing other stuffs at the same moment)

**Comment by xoofx** [comment]
ok build is back to green, sorry for the noisy commits

**Comment by fpetrogalli-arm** [comment]
This LGTM, but I think it is fair to give @diaena some time to review it as she is the one who contributed most (if not all!) the cmake build system. @xoofx is that OK?

**Comment by fpetrogalli-arm** [comment]
Also, as a matter of fact, the convention for writing commit messages needs to be followed only in the merge/squash commit, you don't need to submit each commit in the P in that format. It is fine to undo things, to add new commits on top of each other to address review comments. I am just saying in case, I don't want you to put extra effort in that when it is not needed.

**Comment by xoofx** [comment]
> This LGTM, but I think it is fair to give @diaena some time to review it as she is the one who contributed most (if not all!) the cmake build system. @xoofx is that OK?

Great, thanks! Sure, I will be glad if @diaena can have a look, as long as it won't take too long... 😅 I'm waiting to push the other PRs (MSVC, LLVM Bitcode) and I have to move to other parts on our project (depending on these LLVM/MSVC build), so a context switching here is costly for me.

**Comment by fpetrogalli-arm** [comment]
@xoofx , I appreciate the fact that you need this asap, but I kindly ask you to give @diaena one day to reply. If she doesn't, you can proceed with the squash+merge.

As a practical matter, I think you can submit a new PR with the changes in this PR + the new changes that you need for MSVC. Once #69 is merged, you could rebase those additional PRs to remove the changes of the refactoring.

Thanks!

**Comment by fpetrogalli-arm** [comment]
@xoofx Actually, #69  is just a commit, so yes, it will be very easy to rebase/merge things from additional PRs using it.

**Comment by xoofx** [comment]
Note that my change to the cmake build don't track fine grained dependencies between files (as it was done before). I will make a different PR to fix this, but I believe it is not critical for this PR.

**Comment by fpetrogalli-arm** [comment]
Sorry, I don't understand what you mean with "fine grain dependencies".
________________________________________

**Comment by xoofx** [comment]
> Sorry, I don't understand what you mean with "fine grain dependencies".

Oops, did not explain well. I have added dependencies between targets, but we need some explicit dependencies between files and targets, so that if a file is changing, it should trigger a rebuild. Typically, a change to a file like `sleeflibm.h.org` should make sure that it retrigger the regeneration of the headers, and if not, it should not change. Currently, I don't think that cmake tracks this (like the command `cmake -E copy ${SLEEF_ORG_HEADER} ${SLEEF_GEN_HEADER}`). So everytime to do a `cmake --build ` it will regenerate the files even if there was not any changes.

**Comment by xoofx** [comment]
I have fixed the file dependencies, but only on top of #70 MSVC, as the changes are conflicting to track this with a rebase for both #69 and #70, it's becoming otherwise too laborious. 
I will open a PR for this when both #69 and #70 are merged.

**Comment by fpetrogalli-arm** [comment]
The changes you added to https://github.com/shibatch/sleef/pull/69/commits/14572189a3d4d7942d9803caf58966e254635c3a are breaking the build. Or am I missing something?

**Comment by xoofx** [comment]
> The changes you added to 1457218 are breaking the build. Or am I missing something?

Yeah, sorry for that, I recommit something just after in 797bc37 and it is all green

**Comment by fpetrogalli-arm** [comment]
I see, thanks for explaining. I haven't heard anything back from @diaena. Feel free to squash+merge this patch, it will make things easier.

**Comment by xoofx** [comment]
> I see, thanks for explaining. I haven't heard anything back from @diaena. Feel free to squash+merge this patch, it will make things easier.

**Comment by xoofx** [comment]
> I see, thanks for explaining. I haven't heard anything back from @diaena. Feel free to squash+merge this patch, it will make things easier.

done (and also I have rebased #70 and #72 accordingly)

**Comment by xoofx** [comment]
build is green after squash

**Comment by fpetrogalli-arm** [comment]
You don't need to resubmit the whole branch with the squashed commit. Why don't you use the "squash and merge" button of github? :)

**Comment by fpetrogalli-arm** [comment]
I have merge it for you. Thanks again, this looks very nice.

**Comment by xoofx** [comment]
> I have merge it for you. Thanks again, this looks very nice.

amazing, thanks!  The real one that is correct and nice is the following #72 as it makes the libm build part easier to understand (while still a bit convoluted with the dependencies but hey... 😉  )

**Comment by xoofx** [comment]
> You don't need to resubmit the whole branch with the squashed commit. Why don't you use the "squash and merge" button of github? :)

Because my following PR are depending on it, so I prefer to manage them myself.

**Comment by fpetrogalli-arm** [comment]
Because my following PR are depending on it, so I prefer to manage them myself.

I see, makes sense. Thanks
IMPORTANT NOTICE: The contents of this email and any attachments are confidential and may also be privileged. If you are not the intended recipient, please notify the sender immediately and do not disclose the contents to any other person, use it for any purpose, or store or copy the information in any medium. Thank you.

**Comment by diaena** [comment]
This is coming in a bit late, but it's very good changes.

**Comment by shibatch** [comment]
Welcome back!
I was hoping if you two can collaborate, but it now seems better to think of new things for you to contribute.

**Comment by xoofx** [comment]
> This is coming in a bit late, but it's very good changes.

Thanks @diaena You did the whole initial laborious work! 😉  I wouldn't have considered to take the time to work on this if I had to build the entire cmake infrastructure! (I would have hacked something outside for my very narrow usecase of LLVM bitcode)

**Pull request merged**